File: packages/test-plugin/ios/Sources/TestPlugin/TestPlugin.swift
import Foundation
import Capacitor

/**
 * This file defines the Capacitor bridge for the Test plugin.
 *
 * Documentation reference:
 * https://capacitorjs.com/docs/plugins/ios
 */

@objc(TestPlugin)
public class TestPlugin: CAPPlugin, CAPBridgedPlugin {

    // Configuration instance
    private var config: TestConfig?

    /**
     * Called when the plugin is loaded.
     */
    override public func load() {
        log("Loading plugin")
        self.config = TestConfig(
            config: self.getConfig() as? [String: Any] ?? [:]
        )
    }

    /// Plugin version (Dynamic read from the Bundle Info.plist)
    private var pluginVersion: String {
        let bundle = Bundle(for: type(of: self))
        return bundle.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.1"
    }

    /// The unique identifier for the plugin.
    public let identifier = "TestPlugin"

    /// The name used to reference this plugin in JavaScript.
    public let jsName = "Test"

    /**
     * A list of methods exposed by this plugin. These methods can be called from the JavaScript side.
     * - `echo`: A method that accepts a string and returns the same string.
     * - `getPluginVersion`: A method that returns the version of the plugin.
     */
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    /// An instance of the implementation class that contains the plugin's core functionality.
    private let implementation = Test()

    /**
     * Echoes a string back to JavaScript.
     *
     * @param call Capacitor plugin call containing the `value` parameter.
     */
    @objc func echo(_ call: CAPPluginCall) {
        var value = call.getString("value", "")
        log("Echoing value: \(value)")

        // Append the custom message from the configuration
        if let configMessage = config?.customMessage {
            value = value + configMessage
        }

        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    /**
     * Returns the plugin version.
     */
    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }
}

File: packages/test-plugin/android/src/main/java/com/capkit/test/TestPlugin.kt
package com.capkit.test

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * Capacitor bridge for the Test plugin.
 *
 * This class exposes native Android functionality to JavaScript
 * and delegates all business logic to the implementation class.
 */
@CapacitorPlugin(
  name = "Test",
)
class TestPlugin : Plugin() {
  private lateinit var config: TestConfig

  override fun load() {
    super.load()
    Logger.debug("Loading plugin")
    config = TestConfig(this)
  }

  /** Plugin version (injected by Gradle). */
  private val pluginVersion: String = BuildConfig.PLUGIN_VERSION

  /** Native implementation (business logic). */
  private val implementation = Test()

  /**
   * Echoes a string back to JavaScript.
   *
   * @param call Capacitor plugin call containing the `value` parameter.
   */
  @PluginMethod
  fun echo(call: PluginCall) {
    var value = call.getString("value") ?: ""
    Logger.debug("Echoing value: $value")

    // Append the custom message from the configuration
    value += config.customMessage

    val ret = JSObject()
    ret.put("value", implementation.echo(value))
    call.resolve(ret)
  }

  /**
   * Returns the plugin version.
   */
  @PluginMethod
  fun getPluginVersion(call: PluginCall) {
    val ret = JSObject()
    ret.put("version", pluginVersion)
    call.resolve(ret)
  }
}
